src/rdiffbackup/locations/_repo_shadow.py

```diff
@@ -2306,11 +2306,15 @@ def _restore_orig_regfile(self, rf):
             "Metadata path '{mp}' can only be regular file.".format(
                 mp=rf.metadata_rorp))
         if rf.mirror_rp.isreg():
-            tf = rf.mirror_rp.get_temp_rpath(sibling=True)
-            tf.write_from_fileobj(rf.get_restore_fp())
-            tf.fsync_with_dir()  # make sure tf fully written before move
-            rpath.copy_attribs(rf.metadata_rorp, tf)
-            rpath.rename(tf, rf.mirror_rp)  # move is atomic
+            # Before restoring file from history, let check if the file are already identical.
+            if rf.metadata_rorp.has_sha1() and rf.mirror_rp.getsize() == rf.metadata_rorp.getsize() and rf.metadata_rorp.get_sha1() == hash.compute_sha1(rf.mirror_rp):
```

Comment:
I would prefer to have the size comparison before all the hash related ones, just cosmetic.